### ISSUE TITLE
Detect if Main Viewport is focused

### ISSUE THREAD

**Issue by sammyfreg** [ocornut/imgui#4827]
**Version:** 1.86 WIP (18510)
**Branch:** Docking
**Backend:** Window/DirectX11

I am trying to add partial support of Viewport to NetImgui. At the moment, I deactivate the Viewport support flag when connecting, avoiding issues.

My current idea was to handle the Main Viewport only, leaving the other 'Poped out' windows management to the client, as before.

To achieve this, I was planning to: 
1. Only intercept the Draw Commands of the Main Viewport (seems ok).
2. Only replace the **local inputs** with **remote server input**, when the focus is not on the Main Viewport.

I don't think I can achieve number 2 without some change to the API. 

So far, I have thought of 2 ways of detecting when I should replace the input:
A. Request which window had the focus last frame, and compare its owner ViewportID to the Main ViewportId.
B. See if The Main Viewport is the top Viewport using **LastFrontMostStampCount** (inspired by ShowMetricsWindow), or rely on **PlatformLastFocusedViewportId**.

Point A doesn't exist, and Point B needs access to internal Dear ImGui data.

So, my question was, is there a possible way I missed that I could use, or an API change possible to allow this?

Thank you.

**Comment by ocornut** [comment]
Hello,
I am not exactly sure what's blocking you, For the kind of tech you are developing, I find it reasonable to peek into imgui_internal.h.

You can do like Point A suggest, use `g.NavWindow->VIewport`, note that g.NavWindow may be NULL when no window have focus.

**Comment by sammyfreg** [comment]
I had a mental block on using 'internal' content and needing the extra include specified. But I just remembered that I already had to rely on it for the hook, so I will just use the 'LastFrontMostStampCount' sorting order to infer which is the top most viewport.

Thank you.

**Comment by ocornut** [comment]
Thanks! Will close this for now.

`LastFrontMostStampCount` is a bit of a mess because the relative Z-order of secondary viewports relies on whether they are created as OS child window or not, which depends on the back-end and we have no feedback on that. The TL;DR; is that our knowledge of Z-order is somewhat too broken presently and I think we'll eventually have to find a better solution for this. I'll let you know if there are changes.

**Comment by sammyfreg** [comment]
Thank you.

I was able to rely on `LastFrontMostStampCount` for finding the front most Viewport (on Windows) but encountered some other complications I am not sure can be cleanly solved. As previously mentioned, my first idea to _easily_ for adding Viewport support, was to leave all Viewport management on the Client, and only take over the Main Viewport.

**Problems**
- **Main Window Size** : When NetImgui takes over, it updates the ImGui area size to the Server window size. Being a different value than the local client window size, it confuses the Dear Imgui Window dragging feature (to dock or stop having its own viewport). Haven't investigated it further, but might be possible to restore the Window size after Main Viewport drawing, so this feature keeps works properly.
- **Input management and focus** : I am not sure how I should handle 2 input streams properly. At first, I was thinking of taking over client input only when a secondary Viewport isn't in focus. This somewhat work, but I think it makes for a clunky interaction on the server. What should I do when a local user is using a secondary Viewport and remote user is trying to also use the Main Viewport? We would need to merge keys input (possible) but we can't have 2 focused Windows at the same time...

So, for now I have shifted my focus to adding data compression support (to reduce communication bandwidth). If I decide to go forward with Viewport support, may it would be better to move all of the Viewports to the Server Application too. This would involved some complications to spawn/despawn viewport when connection is established/lost.
